plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/impl/data/ExecuteInsertBatchImpl.java

```diff
@@ -157,56 +161,38 @@ StringBuilder prepareQueryForStatement(
-            if (DBUtils.isPseudoAttribute(attribute) || (!useMultiRowInsert && (!allNulls && DBUtils.isNullValue(attributeValues[i])))) {
-                continue;
-            }
-            if (hasKey) valuesPart.append(","); //$NON-NLS-1$
-            hasKey = true;
+        
+        int usedAttributesArr[] = usedAttributes.stream().mapToInt(Integer::intValue).toArray();
 
-            DBDValueHandler valueHandler = handlers[i];
-            
-            if (valueHandler instanceof DBDValueBinder) {
-                valuesPart.append(((DBDValueBinder) valueHandler).makeQueryBind(attribute, attributeValues[i]));
-            } else {
-                valuesPart.append("?"); //$NON-NLS-1$
-            }
-        }
-        valuesPart.append(")"); //$NON-NLS-1$
-        if (useMultiRowInsert) {
-            for (int i = 0; i < attributeValues.length / attributes.length; i++) {
-                if (i != 0) {
-                    query.append(",");
+        StringJoiner valuesPart = new StringJoiner(","); //$NON-NLS-1$
+        boolean skipBindValues = CommonUtils.toBoolean(options.get(DBSDataManipulator.OPTION_SKIP_BIND_VALUES));
+        for (int i = 0; i < attributeValues.length / attributes.length; i++) {
+            StringJoiner rowValuesPart = new StringJoiner(",", "(", ")"); //$NON-NLS-1$
+            for (int j : usedAttributesArr) {
+                DBSAttributeBase attribute = attributes[j];
+                DBDValueHandler valueHandler = handlers[j];
+                int k = i * attributes.length + j;
+                if (valueHandler instanceof DBDValueBinder) {
+                    rowValuesPart.add(((DBDValueBinder) valueHandler).makeQueryBind(attribute, attributeValues[k]));
+                } else {
+                    if (skipBindValues) {
+                        rowValuesPart.add(SQLUtils.convertValueToSQL(session.getDataSource(), attribute, valueHandler, attributeValues[k], DBDDisplayFormat.NATIVE));
+                    } else {
+                        rowValuesPart.add("?"); //$NON-NLS-1$
+                    }
                 }
-                query.append(valuesPart);
             }
-        } else {
-            query.append(valuesPart);
-        }
-
-        if (skipBindValues) {
-            // Here we replace all binding marks (?) with real values
-            int curPos = 0;
-            for (int i = 0; i < attributeValues.length; i++) {
-                String sqlValue = SQLUtils.convertValueToSQL(
-                    session.getDataSource(),
-                    attributes[i % attributes.length],
-                    handlers[i % handlers.length],
-                    attributeValues[i],
-                    DBDDisplayFormat.NATIVE);
-
-                int qmPos = query.indexOf("?", curPos);
-                query.replace(qmPos, qmPos + 1, sqlValue);
+            valuesPart.add(rowValuesPart.toString());
+            if (!useMultiRowInsert) {
+                break;
             }
```

Comment:
If multi-row insert is disabled, values and attributes should have same length. I think this statement can be safely avoided by adding an assertion in the beginning of the method (along with `handlers.length == attributes.length` mentioned some time ago), just to be extra sure. 

You can use `org.eclipse.core.runtime.Assert#isLegal` for this purpose.

Reply:
Done